### ISSUE TITLE
[Sprint] numpy & matplotlib notes

### ISSUE THREAD

**Issue by eah13** [skulpt/skulpt#399]
These are both massive modules.  The notes below may help a developer prioritize.

Matplotlib stuff from Trinket error logs:
- [ ] matplotlib.figure()
- [ ] matplotlib.pyplot.waitforbuttonpress()
- [ ] matplotlib.pyplot.scatter()
- [ ] matplotlib.pyplot.rcdefaults() (doc)
- [ ] xlim and ylim (parameters of plots I think?)

Numpy stuff from Trinket error logs:
- [ ] numpy.random
- [ ] numpy.random.random

Bonus: symlink the relevant pylab modules

**Comment by delijati** [comment]
Some parts are already implemented here https://github.com/tum-vision/autonavx_skulpt

**Comment by albertjan** [comment]
I love stuff like this popping up! Thanks @delijati for pointing us to it.

**Comment by delijati** [comment]
Yeah its great to see python in the browser. My Girlfriend used this in a mooc http://www.codeskulptor.org. What i saw they added a SimpleGUI framework to build nice litte apps.

**Comment by albertjan** [comment]
@waywaaard did the guys over the @tum-vision implement any numpy or mathplotlib stuff you didn't? Might be interesting to look at it.

**Comment by waywaaard** [comment]
They developed a completely new ndarray version and added a few other functions. I need to have a closer look at it.

I think my implementation of ndarray is completer and has better slicing support. However, have they implemented parts of the random module, that we could use/merge.

**Comment by eah13** [comment]
Update: This work is being done in https://github.com/waywaaard/skulpt_numpy and https://github.com/waywaaard/skulpt_matplotlib  Please join us in those repos if you're interested in contributing.

**Comment by JdeH** [comment]
There's also a port of ndarray (real and complex), * / + - @, inverse, FFT, IFFT, FFT2, IFFT2, hstack vstack hsplit vspit at

https://github.com/JdeH/Numscrypt

It was written for Transcrypt, but since it's Python-only, I guess it probably could be used for Skulpt as well. It has been highly simplified, omitting reshaping and views, and supporting at most 2 dimensions. This means that it can be rather fast, also since it uses typed array. While Numscrypt itself will remain geared towards Transcrypt, feel free to fork from it and use it as a basis. I think the basis of the design isn't too bad.

Kind regards
Jacques de Hooge

**Comment by waywaaard** [comment]
Thanks for the hint. Yeah, the current module has some problems with slicing.

**Comment by JdeH** [comment]
The strides / slicing etc. code is still in the GitHub history of Numscrypt. It did the right thing AFAIK, but it was far to slow in JS. I did some searching and it seems that ndarrays with dimension > 2 are rarely used so they were left out as well.

**Comment by bnmnetp** [comment]
@ebertmi is the matplotlib repo in your home the most up to date version?  I am doing a bunch of content updates this summer and would like to incorporate your work on matplotlib and numpy.

Thanks!

Brad

**Comment by psychemedia** [comment]
I'm not sure how useful this is, but I note that the BlockPy team's fork of Skulpt includes a [matplotlib.pylot module](https://github.com/blockpy-edu/skulpt/tree/master/src/lib/matplotlib).
